lib/reek/smells/nested_iterators.rb

```diff
@@ -79,31 +83,33 @@ def find_deepest_iterator(ctx)
       #   foo.each() do ... end
       # this will end up as:
       #
-      # "foo.each() do ... end" -> the iterator below
-      # "each()"                -> the "call" below
-      # "do ... end"            -> the "block" below
-      #
-      # @param parent [AST::Node] The parent iterator
+      # "foo.each() do ... end" -> one of the :block nodes
+      # "each()"                -> the node's "call"
+      # "do ... end"            -> the node's "block"
       #
       # @param exp [AST::Node]
       #   The given expression to analyze.
-      #   Will be nil on empty blocks so we'll return just the parent iterator
       #
       # @param depth [Integer]
       #
       # @return [Array<Iterator>]
       #
-      def scout(parent:, exp:, depth:)
-        return [Iterator.new(parent, depth)] unless exp
-        iterators = exp.find_nodes([:block])
-        return [Iterator.new(parent, depth)] if iterators.empty?
-        iterators.map do |iterator|
+      # :reek:TooManyStatements: { max_statements: 6 }
+      def scout(exp:, depth:)
+        return [] unless exp
+        exp.find_nodes([:block]).flat_map do |iterator|
+          new_depth = increment_depth(iterator, depth)
           # 1st case: we recurse down the given block of the iterator. In this case
           # we need to check if we should increment the depth.
           # 2nd case: we recurse down the associated call of the iterator. In this case
           # the depth stays the same.
-          scout(parent: iterator, exp: iterator.block, depth: increment_depth(iterator, depth)) +
-            scout(parent: iterator, exp: iterator.call, depth: depth)
+          nested_iterators = scout(exp: iterator.block, depth: new_depth) +
+            scout(exp: iterator.call, depth: depth)
+          if nested_iterators.empty?
+            Iterator.new(iterator, new_depth)
+          else
+            nested_iterators
+          end
```

Comment:
Hmmm, just a small thing: Now it's not tail-recursive anymore. Given the non-performance critical domain of Reek this is not a big problem, but I wanted at least to point it out.